os/kernel/debug/mem_leak_checker.c

```diff
@@ -0,0 +1,407 @@
+				(*broken_cnt)++;
+				continue;
+			}
+			node_size = node->size;
+			if ((unsigned long)node + (unsigned long)SIZEOF_MM_ALLOCNODE == (unsigned long)node_info || 
+					(unsigned long)node + (unsigned long)SIZEOF_MM_ALLOCNODE == (unsigned long)hash_table) {
+				continue;
+			}
+			/* Check if the node corresponds to an allocated memory chunk */
+			if ((node->preceding & MM_ALLOC_BIT) != 0) {
+				node_info[*leak_cnt].node = node;
+				node_info[*leak_cnt].next = NULL;
+				node->reserved = MEM_LEAK;
+				add_hash(*leak_cnt);
+				(*leak_cnt)++;
+			}
+		}
+	}
+	mm_givesemaphore((struct mm_heap_s *)leak_checker.heap);
+}
+
+static void search_addr(void *start_addr, void *end_addr, int *leak_cnt)
+{
+	/* This function traverse the memory from start_addr to end_addr for comparing the address based on hash table. */
+	void *leak_chk;
+	/* Not to access over its region, subtract 0x04 from the end of the address. */
+	for (leak_chk = start_addr; leak_chk < end_addr - MEM_ACCESS_UNIT; leak_chk++) {
+		if (leak_chk == (void *)&node_info) {
+			/* Skip to check hash table */
+			leak_chk += (sizeof(struct alloc_node_info_s) * MAX_ALLOC_COUNT);
+			leak_chk--;
+			continue;
+		}
+		if (search_hash(*(unsigned long volatile *)leak_chk - (unsigned long)SIZEOF_MM_ALLOCNODE)) {
+			(*leak_cnt)--;
+		}
+	}
+}
+
+static void heap_check(int *leak_cnt)
+{
+	void *leak_chk;
+	struct mm_allocnode_s *visit_node;
+	void *exclude_top;
+	void *exclude_bottom;
+
+	exclude_top = leak_checker.stack_top;
+	exclude_bottom = leak_checker.stack_bottom;
+
+#if CONFIG_KMM_REGIONS > 1
+	int region;
+#else
+#define region 0
+#endif
+
+	/* Visit each region */
+
+#if CONFIG_KMM_REGIONS > 1
+	for (region = 0; region < leak_checker.regions; region++)
+#endif
+	{
+		for (visit_node = leak_checker.heap_start[region]; visit_node < leak_checker.heap_end[region]; visit_node = (struct mm_allocnode_s *)((char *)visit_node + visit_node->size)) {
+			if ((visit_node->preceding & MM_ALLOC_BIT) != 0) {
+				for (leak_chk = (void *)visit_node; leak_chk < (void *)(((char *)visit_node) + visit_node->size); leak_chk++) {
+					if ((leak_chk >= exclude_bottom && leak_chk <= exclude_top)) {
+						continue;
+					}
+					if (search_hash(*(unsigned long volatile *)leak_chk - (unsigned long)SIZEOF_MM_ALLOCNODE)) {
+						(*leak_cnt)--;
+					}
+				}
+			}
+		}
+	}
+}
+
+int checkerpid;
+
+static void init_mem_leak_checker(int checker_pid, char *bin_name);
+
+static void ram_check(char *bin_name, int *leak_cnt, uint32_t *bin_text_addr)
+{
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	bin_addr_info_t *info;
+	int bin_idx;
+
+	info = (bin_addr_info_t *)get_bin_addr_list();
+	for (bin_idx = 0; bin_idx <= CONFIG_NUM_APPS; bin_idx++) {
+		if (strncmp(BIN_NAME(bin_idx), bin_name, strlen(bin_name)) == 0) {
+			/* For calculating the owner from binary, save the text section address. */
+			*bin_text_addr = info[bin_idx].text_addr;
+			break;
+		}
+	}
+#endif
+	/* Visit all the data regions
+	 */
+	int i;
```

Comment:
As you can see above (bin_idx), we want to use readable, meaningful variable name. Could you rename this?
If we don't, we don't need to use two variables for for loop.